File: examples/python/synchro-barrier/synchro-barrier.py
# Copyright (c) 2010-2022. The SimGrid Team. All rights reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the license (GNU LGPL) which comes with this package.

from argparse import ArgumentParser
import sys

from simgrid import Actor, Barrier, Engine, Host, this_actor


def create_parser() -> ArgumentParser:
    parser = ArgumentParser()
    parser.add_argument(
        '--platform',
        type=str,
        required=True,
        help='path to the platform description'
    )
    parser.add_argument(
        "--actors",
        type=int,
        required=True,
        help="Number of actors to start"
    )
    return parser


def worker(barrier: Barrier):
    """ Wait on the barrier and exits.
    :param barrier: Barrier to be awaited
    """
    this_actor.info(f"Waiting on the barrier")
    barrier.wait()
    this_actor.info("Bye")


def master(actor_count: int):
    """ Create barrier with `actor_count` expected actors, spawns `actor_count - 1` workers, then wait on the barrier
    and exits.
    :param actor_count: Spawn actor_count-1 workers and do a barrier with them
    """
    barrier = Barrier(actor_count)
    workers_count = actor_count - 1
    this_actor.info(f"Spawning {workers_count} workers")
    for i in range(workers_count):
        Actor.create(f"worker-{i}", Host.by_name("Jupiter"), worker, barrier)
    this_actor.info(f"Waiting on the barrier")
    barrier.wait()
    this_actor.info("Bye")


def main():
    settings = create_parser().parse_known_args()[0]
    if settings.actors < 1:
        raise ValueError("--actors must be greater than 0")
    e = Engine(sys.argv)
    e.load_platform(settings.platform)
    Actor.create("master", Host.by_name("Tremblay"), master, settings.actors)
    e.run()


if __name__ == "__main__":
    main()

File: examples/java/synchro_barrier/synchro_barrier.java
/* Copyright (c) 2006-2025. The SimGrid Team. All rights reserved.          */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

import org.simgrid.s4u.*;

/// Wait on the barrier then leave
class worker extends Actor {
  Barrier barrier;
  public worker(Barrier barrier) { this.barrier = barrier; }
  public void run()
  {
    Engine.info("Waiting on the barrier");
    if (barrier.await())
      Engine.info("Bye from the last to enter");
    else
      Engine.info("Bye");
  }
}

/// Spawn actor_count-1 workers and do a barrier with them
class master extends Actor {
  int actor_count;
  public master(int actor_count) { this.actor_count = actor_count; }
  public void run()
  {
    Engine e = this.get_engine();

    Barrier barrier = Barrier.create(actor_count);

    Engine.info("Spawning %d workers", actor_count - 1);
    for (int i = 0; i < actor_count - 1; i++)
      e.host_by_name("Jupiter").add_actor("worker", new worker(barrier));

    Engine.info("Waiting on the barrier");
    if (barrier.await())
      Engine.info("Bye from the last to enter");
    else
      Engine.info("Bye");
  }
}
public class synchro_barrier {
  public static void main(String[] args)
  {
    Engine e = new Engine(args);

    // Parameter: Number of actors in the barrier
    if (args.length < 1)
      Engine.die("Usage: synchro_barrier <actor-count>\n");

    int actor_count = Integer.valueOf(args[0]);
    if (actor_count <= 0)
      Engine.die("<actor-count> must be greater than 0");

    e.load_platform(args.length >= 2 ? args[1] : "../platforms/two_hosts.xml");
    e.host_by_name("Tremblay").add_actor("master", new master(actor_count));
    e.run();

    // The following call is useless in your code, but our continuous integration uses it to track memleaks
    e.force_garbage_collection();
  }
}
